internal/plugin/connectors/tcp/mssql/mock/options.go

```diff
@@ -0,0 +1,103 @@
+package mock
+
+import (
+	"context"
+	"net"
+
+	logmock "github.com/cyberark/secretless-broker/pkg/secretless/log/mock"
+
+	"github.com/cyberark/secretless-broker/internal/plugin/connectors/tcp/mssql/types"
+	mssql "github.com/denisenkom/go-mssqldb"
+)
+
+// MSSQLConnectorCtorOptions represents options when creating a MSSQLConnectorCtor. This
+// makes it possible to customize some key values in a Connect double:
+// func Connect(...) (net.Conn, error)
+type MSSQLConnectorCtorOptions struct {
+	// Backend connection returned from the call to Connect
+	BackendConn net.Conn
+	// Error returned from the call to Connect
+	Err error
+	// PreloginResponse from the server passed to interceptor during Connect
+	ServerPreloginResponse map[uint8][]byte
+	// Pointer to unload client login request to, taken from interceptor during Connect
+	ClientLoginRequestPtr **mssql.LoginRequest
+	// LoginResponse from the server passed to interceptor during Connect
+	ServerLoginResponse *mssql.LoginResponse
+}
+
+// MSSQLConnectorCtorOption is the 'functional option' complement to
+// MSSQLConnectorCtorOptions.
+type MSSQLConnectorCtorOption func(*MSSQLConnectorCtorOptions)
+
+// MSSQLConnectorCtor is a 'functional option' of types.ConnectorOption. It generates
```

Comment:
I thought about this for a long time, and wrote up a whole long thing about how to make it clearer, improve names, etc, but at the end I threw it all away because I think this is a mistake.

Even being very familiar with these patterns, it took me too long to understand the code, and there are too many types to keep in your head.  It reminds of:

"All problems in computer science can be solved by another level of indirection, except too much indirection."

This is something I struggle with all the time and think about a lot.  For my money, though, the nested functional options veers clearly into unnecessary abstraction territory here.  If I were to revisit this code in 6mos, I would regret having written it that way.

And if you look at [Rob's original article](https://commandcenter.blogspot.com/2014/01/self-referential-functions-and-design.html) on this, the problem it's trying to solve is to make an API growable and flexible.  These conditions don't apply here.  This mock code is private and unlikely to grow. (It's even debatable if the main usage of it in the connector is justified, but that part reads clearly to me and I think I like it.)

I would make these changes:

1. Rename `mock.MSSQLConnectorCtorOptions` to `MSSQLConnectorCtor`
1. Rename what is currently `MSSQLConnectorCtor` to `MSSQLConnectorCtorOpt`.  It returns an option after all, per the first line of your comment.
2. Rewrite usages such as the following:

```go
connector := NewSingleUseConnectorWithOptions(
  mock.DefaultConnectorOptions(),
  mock.MSSQLConnectorCtor(
    func(options *mock.MSSQLConnectorCtorOptions) {
      options.BackendConn = expectedBackendConn
    },
  ),
)
```
to something like:

```go
connector := NewSingleUseConnectorWithOptions(
  mock.DefaultConnectorOptions(),
  mock.MSSQLConnectorCtorOpt( // returns a connector option
    mock.MSSQLConnectorCtor{
      BackendConn: expectedBackendConn
    }
  ),
)
```
You can fill in the details so that default values are taken care of as elegantly as they are with the functional options pattern.